README.md
# CQ-picfinder-robot

# 已经连node-v13.9.0-windows-x64一块打包了

# 如果是使用高版本的nodejs，需要删掉package-lock.json，再自己安装一次 sqlite sqlite3 canvas

# 当前改为适配go-cqhttp https://github.com/Mrs4s/go-cqhttp/ 

# 连接方式改为 HTTP POST多点上报 https://github.com/howmanybots/cqhttp-node-sdk

npm i sqlite3

https://www.npmjs.com/package/sqlite

https://www.npmjs.com/package/sqlite3

https://www.npmjs.com/package/canvas

当前跟进原作者版本：v2.18.1

未跟进内容有：

v2.13.2
- 弃用 mysql，仅使用 sqlite，配置项转移
v2.14.2
- 自定义每日资料卡点赞名单
- 配置项变更
v2.15.0 
- 配置项变更
  - M `picfinder` -> `bot`，会自动迁移，无需手动更改
v2.15.1
- 为`config.json`增加`$schema`
v2.15.2
- 搜图参数及图库关键字中的`book`修改为`doujin`，但`book`依然可用
v2.16.1
- 每日任务
- debug（调试）模式下长消息截断
v2.17.0
- 支持[配置热重载](https://github.com/Tsuk1ko/cq-picsearcher-bot/wiki/%E5%A6%82%E4%BD%95%E9%A3%9F%E7%94%A8#%E9%85%8D%E7%BD%AE%E7%83%AD%E9%87%8D%E8%BD%BD)
- 改进定时提醒的逻辑
- 修复机器人手动入群后没有文字反馈的问题
v2.17.0
- fix: 发送两次已上线
v2.18.0
- 增加[语言库](https://github.com/Tsuk1ko/cq-picsearcher-bot/wiki/%E9%99%84%E5%8A%A0%E5%8A%9F%E8%83%BD#%E8%AF%AD%E8%A8%80%E5%BA%93%E8%87%AA%E5%8A%A8%E5%9B%9E%E5%A4%8D)功能（自动回复）
- 配置项变更
  - A `bot.corpus`
v2.18.1
- 修复消息群发失效 ([#101](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/101))
- chore: 改用全局函数

1、强化了b站视频av/bv号解析功能，改了下解析b站分享小程序判断

2、强化搜图防御（可以让指定群不能搜图，相似度低时，saucenao或WhatAnime均会隐藏简略图（90%以上相似度才不隐藏，r18结果大概也都会隐藏（没彻底测试过）），只有指定群和指导好友QQ可以用a2d搜图，因为a2d很难防御）

3、签到改为出图片功能，图片名字要求为纯数字.jpg，例如：0.jpg,1.jpg...

待补充...


这是一个以 Nodejs 编写的酷Q机器人插件，用于搜图、搜番、搜本子，并夹带了许多娱乐向功能（。）

目前支持：

- [saucenao](https://saucenao.com)
- [WhatAnime](https://trace.moe)
- [ascii2d](https://ascii2d.net)

附加功能：

- 复读
- setu
- OCR
- 明日方舟公开招募计算
- 定时提醒

详细说明请移步 [Wiki](https://github.com/Tsuk1ko/CQ-picfinder-robot/wiki)

## TODO

重构为更加模块化的一个机器人框架，以支持自定义插件

# 其它以CQ-picfinder-robot为基础的魔改版
- https://github.com/Ninzore/Wecab

# 安装模块失败的解决方法
* https://github.com/Tsuk1ko/CQ-picfinder-robot/issues/42#issuecomment-572051285
## 1
* https://npm.taobao.org/
* npm config set registry https://registry.npm.taobao.org
* npm config get registry
## 2
单纯挂的系统代理是没用的，因为命令行并不走系统代理，你需要用 [sstap](https://github.com/FQrabbit/SSTap-Rule) 之类的工具代理，或者尝试按下面的方法为 npm 设置代理
以本地端口为1080的小飞机为例
* npm config set proxy http://127.0.0.1:1080
* npm config set https-proxy http://127.0.0.1:1080
# 如果要取消
* npm config delete proxy
* npm config delete https-proxy


CHANGELOG.md
# 更新日志

## 2020

### 09-13 v2.18.1

- 修复消息群发失效 ([#101](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/101))

### 09-12 v2.18.0

- 增加[语言库](https://github.com/Tsuk1ko/cq-picsearcher-bot/wiki/%E9%99%84%E5%8A%A0%E5%8A%9F%E8%83%BD#%E8%AF%AD%E8%A8%80%E5%BA%93%E8%87%AA%E5%8A%A8%E5%9B%9E%E5%A4%8D)功能（自动回复）
- 配置项变更
  - A `bot.corpus`

### 09-11 v2.17.1

- 支持`[CQ:json]` ([#100](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/100))

### 09-05 v2.17.0

- 支持[配置热重载](https://github.com/Tsuk1ko/cq-picsearcher-bot/wiki/%E5%A6%82%E4%BD%95%E9%A3%9F%E7%94%A8#%E9%85%8D%E7%BD%AE%E7%83%AD%E9%87%8D%E8%BD%BD)
- 改进定时提醒的逻辑
- 修复机器人手动入群后没有文字反馈的问题

### 09-05 v2.16.1

- 修复 JSON 转义问题导致的哔哩哔哩小程序识别错误 ([#96](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/96))
- 改进 nHentai 搜索

### 08-27 v2.16.0

- 增加搜图结果发送缩略图相关的详细设置 ([#90](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/92))
- 配置项变更
  - M `bot.saucenaoHideImgWhenLowAcc` -> `bot.hideImgWhenLowAcc`，会自动迁移，无需手动更改
  - A `bot.hideImg`
  - A `bot.hideImgWhenWhatanimeR18`

### 08-17 v2.15.4

- 修复定时提醒 interval 超出 32 位有符号整数导致的刷屏问题 ([#90](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/90))

### 08-16 v2.15.3

- 修复反 Bilibili 小程序会响应动态的小程序分享的问题 ([#89](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/89))

### 08-15 v2.15.2

- 修复反 Bilibili 小程序的防刷屏逻辑问题 ([#87](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/87))
- 搜图参数及图库关键字中的`book`修改为`doujin`，但`book`依然可用

### 08-14 v2.15.1

- 为`config.json`增加`$schema`
- 恢复群发消息功能 ([#86](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/86))

### 08-13 v2.15.0

- 完全恢复转义，需使用 go-cqhttp v0.9.18 及以上版本
- 在群内发送搜图结果将会采用回复的形式 ([#84](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/84))
- 启动时会检查配置文件是否存在以及 JSON 合法性
- 可独立开关私聊和群组消息的监听
- 配置项变更
  - M `picfinder` -> `bot`，会自动迁移，无需手动更改
  - A `bot.enablePM`
  - A `bot.enableGM`

### 08-10 v2.14.3

- 恢复部分转义，需使用 go-cqhttp v0.9.16 及以上版本

### 08-07 v2.14.2

- 因 go-cqhttp 尚未支持转义 [Mrs4s/go-cqhttp#9](https://github.com/Mrs4s/go-cqhttp/issues/9)，因此暂时禁用了消息的转义，待其修复后需要更新 go-cqhttp 和本项目

### 08-07 v2.14.1

- 之前忘记删除签到（点赞）相关功能代码了

### 08-05 v2.14.0 **R.I.P. CoolQ**

- 目前决定专注于适配 [go-cqhttp](https://github.com/Mrs4s/go-cqhttp)，其余方案暂不考虑
  - mirai-native + cq-http 经测试仍然有很多问题且部署麻烦，因此放弃
  - 如果发现某些功能运作不正常或与原先表现不一致，可提 issue 向我反馈
- 请参考 wiki 进行迁移或部署，另外，配置文件结构有些许变动（主要是 node-cq-websocket 部分），请注意修改
- mirai 不支持点赞，自动点赞功能及相关配置项已被删除

### 07-30 v2.13.5

- 自定义每日资料卡点赞名单
- 配置项变更
  - A `bot.dailyLike`

### 07-24 v2.13.4

- 反 Bilibili 小程序功能在 3 分钟内将不会重复解析同一视频链接，以防刷屏

### 07-06 v2.13.3

- 增加 setu API 超额时的自定义回复
- 配置项变更
  - A `bot.replys.setuQuotaExceeded`

### 07-02 v2.13.2

- 修复定时提醒的逻辑错误

### 06-27 v2.13.1

- 修复搜图缓存没有正常运作的问题

### 06-27 v2.13.0

- 修复提醒功能失效问题 ([#75](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/75))
- 弃用 mysql，仅使用 sqlite，配置项转移
- 配置项变更
  - D `mysql`
  - A `bot.cache`
  - `mysql.enable` -> `bot.cache.enable`
  - `mysql.expire` -> `bot.cache.expire`

### 5-10 v2.12.6

- 修复一个 bug

### 5-10 v2.12.5

- 修复方舟公招数据更新问题

### 5-05 v2.12.4

- 增加私聊回复群聊中搜图结果的功能 ([#60](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/60))
- 配置项变更
  - A `bot.pmSearchResult`

### 5-01 v2.12.3

- 反哔哩哔哩小程序不支持番剧链接，将尽可能忽略番剧链接 ([#59](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/59))

### 4-29 v2.12.2

- 更新方舟公招数据来源

### 4-29 v2.12.1

- 修正 debug 逻辑 ([#58](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/58))

### 4-27 v2.12.0

- 增加“反哔哩哔哩小程序”功能，鼓励发链接，发链接时会自动获取视频信息并发送，详情看 wiki 配置说明及附加功能
- 配置项变更
  - A `bot.antiBiliMiniApp`

### 4-24 v2.11.14

- 更改公开招募计算器触发词，不再需要`--akhr`，改为包含`akhr`或`公招`一词即可

### 4-11 v2.11.13

- 修复 danbooru 获取原图来源问题

### 3-18 v2.11.12

- 修复 whatanime 错误 ([#54](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/54))
- 改进错误输出

### 3-18 v2.11.11

- 修复反和谐生成图片过大问题 ([#53](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/53))

### 3-17 v2.11.10

- 修复定时提醒功能判断分钟级间隔有误的问题

### 3-14 v2.11.9

- 更换 akhr 数据地址 ([#49](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/49))
- 增加 whatanime 的 token 设置
- 配置项变更
  - A `whatanimeToken`

### 3-9 v2.11.8

- 增加 setu 的 apikey 设置
- 配置项变更
  - A `bot.setu.apikey`

### 2-21 v2.11.7

- 修复通用处理完成后未停止事件传播的问题 ([#36](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/36))

### 2-18 v2.11.6

- WhatAnime 使用官方提供的 API

### 2-03 v2.11.5

- 增加 SauceNao 低相似度值自定义配置
- 增加“SauceNao 结果相似度过低时结果缩略图的替代文字”的配置
- 配置项变更
  - A `bot.saucenaoLowAcc`
  - A `bot.replys.lowAccImgPlaceholder`

### 2-01 v2.11.4

- 增加“SauceNao 结果相似度过低时隐藏结果缩略图”的配置
- 配置项变更
  - A `bot.saucenaoHideImgWhenLowAcc`

### 1-29 v2.11.3

- 增加对`http://www.pixiv.net/(artworks|users)/[0-9]+`链接的短缩

### 1-21 v2.11.2

- 增加配置项用于控制是否在 saucenao 结果低相似度或配额耗尽时使用 ascii2d
- 配置项变更
  - A `bot.useAscii2dWhenQuotaExcess`
  - A `bot.useAscii2dWhenLowAcc`

### 1-15 v2.11.1

- 因酷Q不支持本地发送大于 4M 的图片，因此开启反和谐后如果没有开启 size1200 并且原图大小超过 3M，将会自动使用 size1200 ([#40](https://github.com/Tsuk1ko/cq-picsearcher-bot/issues/40))

## 2019

### 12-18 v2.11.0

- 当 ascii2d 失败时返回错误信息
- 支持自定义 ascii2d 的域名
- saucenao, whatanime, ascii2d 的自定义域名支持带上协议，即支持以下写法
  - `example.com`：将会使用`http://example.com`；特殊地，上面三者的官方域名将会使用 https
  - `http://example.com`或`https://example.com`
- 支持[群发消息](https://github.com/Tsuk1ko/cq-picsearcher-bot/wiki/%E5%A6%82%E4%BD%95%E9%A3%9F%E7%94%A8#%E7%BE%A4%E5%8F%91%E6%B6%88%E6%81%AF)
- 配置项变更
  - A `ascii2dHost`

### 12-09 v2.10.1

- 增大 setu 反和谐力度
- 支持获取 yande.re 结果的原出处
- 增加`--help`,`--about`,`--version`命令

### 12-02 v2.10.0

- setu 反和谐
- 配置项变更
  - A `bot.setu.antiShielding`

### 11-05 v2.9.5

- 在 ascii2d 搜索失败时返回失败提示语 #31

### 10-29 v2.9.4

- 使用 named-regexp-groups 模块以解决某些 node 版本莫名其妙无法使用命名正则表达式捕获组的问题
- 搜图错误时的回复增加了 saucenao host index

### 10-25 v2.9.3

- 支持发送 master1200 大小的 setu 以改善小水管或国内机器发图速度
- 配置项变更
  - A `bot.setu.size1200`

### 10-22 v2.9.2

- 修复 admin 搜图时的记录问题
- 修复 npm 脚本错误
- 改善 setu 正则表达式

### 10-15 v2.9.1

- 增加 pm2 配置文件，目前可直接使用`pm2 start|stop|restart|logs`等命令控制
- 增加按关键词发 setu 以及 r18 setu 功能，若从旧版本升级，请参考 wiki 中 setu 功能说明进行设置
- 配置项变更（重要）
  - A `bot.setu.r18OnlyInWhite`
  - M `bot.regs.setu`

### 08-21 v2.8.0

- 增加对提醒功能最小提醒间隔的限制，新增配置项支持限制使用场景

modules/saucenao.js
import {
    get
} from './axiosProxy';
import nhentai from './nhentai';
import getSource from './getSource';
import CQ from './CQcode';
import config from './config';
import shorten from './urlShorten/is.gd';
import {
    parse
} from 'url';
import logger2 from './logger2';
import pixivShorten from './urlShorten/pixiv';
//import logError from './logError';

const hosts = config.saucenaoHost;
let hostsI = 0;

const snDB = {
    all: 999,
    pixiv: 5,
    danbooru: 9,
    book: 18,
    anime: 21,
};

const exts = {
    j: 'jpg',
    p: 'png',
    g: 'gif',
};

const saucenaoApiKeyAddition = config.saucenaoApiKey ? {
    api_key: config.saucenaoApiKey
} : {};

/**
 * saucenao搜索
 *
 * @param {string} imgURL 图片地址
 * @param {string} db 搜索库
 * @param {boolean} [debug=false] 是否调试
 * @returns Promise 返回消息、返回提示
 */
async function doSearch(imgURL, db, debug = false, whitegroup, whiteqq) {
    const hostIndex = hostsI++ % hosts.length; //决定当前使用的host
    let warnMsg = ''; //返回提示
    let msg = config.picfinder.replys.failed; //返回消息
    let success = false;
    let lowAcc = false;
    let excess = false;

    await getSearchResult(hosts[hostIndex], imgURL, db)
        .then(async ret => {
            const data = ret.data;

            //如果是调试模式
            if (debug) {
                logger2.info(`\n${getTime()},saucenao[${hostIndex}]: ${hosts[hostIndex]}`);
                logger2.info(JSON.stringify(data));
                //console.log(`\n[debug] saucenao[${hostIndex}]: ${hosts[hostIndex]}`);
                //console.log(JSON.stringify(data));
            }

            //确保回应正确
            if (data.results && data.results.length > 0) {
                let {
                    header: {
                        short_remaining, //短时剩余
                        long_remaining, //长时剩余
                        similarity, //相似度
                        thumbnail, //缩略图
                    },
                    data: {
                        ext_urls,
                        title, //标题
                        member_name, //作者
                        member_id, //可能 pixiv uid
                        eng_name, //本子名
                        jp_name, //本子名
                    },
                } = data.results[0];

                let url = ''; //结果链接
                let source = null;
                if (ext_urls) {
                    url = ext_urls[0];
                    //如果结果有多个，优先取danbooru
                    for (let i = 1; i < ext_urls.length; i++) {
                        if (ext_urls[i].indexOf('danbooru') !== -1) url = ext_urls[i];
                    }
                    url = url.replace('http://', 'https://');
                    //若为danbooru则获取来源
                    source = await getSource(url).catch(() => null);
                }

                if (!title) title = url.indexOf('anidb.net') === -1 ? ' 搜索结果' : ' AniDB';

                let bookName = jp_name || eng_name; //本子名

                if (member_name && member_name.length > 0) title = `\n「${title}」/「${member_name}」`;

                //剩余搜图次数
                if (long_remaining < 20) warnMsg += CQ.escape(`\nsaucenao[${hostIndex}]：注意，24h内搜图次数仅剩${long_remaining}次\n`);
                else if (short_remaining < 5) warnMsg += CQ.escape(`\nsaucenao[${hostIndex}]：注意，30s内搜图次数仅剩${short_remaining}次\n`);
                //相似度
                if (similarity < config.picfinder.saucenaoLowAcc) {
                    lowAcc = true;
                    warnMsg += CQ.escape(`\n相似度[${similarity}%]过低，如果这不是你要找的图，那么可能：确实找不到此图/图为原图的局部图/图清晰度太低/搜索引擎尚未同步新图\n`);
                    if (config.picfinder.useAscii2dWhenLowAcc && (db == snDB.all || db == snDB.pixiv) && (whitegroup == true || whiteqq == true)) {
                        warnMsg += '自动使用 ascii2d 进行搜索\n';
                    }
                    if (config.picfinder.saucenaoHideImgWhenLowAcc) thumbnail = null;
                }

                //回复的消息
                msg = await getShareText({
                    url,
                    title: `SauceNAO [${similarity}%]${title}`,
                    thumbnail,
                    author_url: member_id && url.indexOf('pixiv.net') >= 0 ? `https://pixiv.net/u/${member_id}${(thumbnail != null ? '\nLook: https://pixivic.com/artist/' + member_id : '')}` : null,
                    source,
                });

                success = true;

                //如果是本子
                if (bookName) {
                    bookName = bookName.replace('(English)', '');
                    const book = await nhentai(bookName).catch(e => {
                        logger2.error(`${getTime()}[error] nhentai:` + e);
                        //console.error(`${new Date().toLocaleString()} [error] nhentai`);
                        //console.error(e);
                        return false;
                    });
                    //有本子搜索结果的话
                    if (book && config.picfinder.saucenaoHideImgWhenLowAcc == false) {
                        thumbnail = `https://t.nhentai.net/galleries/${book.media_id}/cover.${exts[book.images.thumbnail.t]}`;
                        url = `https://nhentai.net/g/${book.id}/`;
                    } else if (config.picfinder.saucenaoHideImgWhenLowAcc == false) {
                        success = false;
                        warnMsg +=
                            '没有在 nhentai 找到对应的本子，或者可能是此 query 因 bug 而无法在 nhentai 中获得搜索结果 _(:3」∠)_\n';
                    }
                    msg = await getShareText({
                        url,
                        title: `[${similarity}%] ${bookName}`
                            /*,
                                                    thumbnail,*/ //禁止发本子简略图
                    });
                }

                //处理返回提示
                if (warnMsg.length > 0) warnMsg = warnMsg.trim();
            } else if (data.header.message) {
                switch (data.header.message) {
                    case 'Specified file no longer exists on the remote server!':
                        msg = '该图片已过期，请尝试二次截图后发送';
                        break;

                    case 'Problem with remote server...':
                        msg = `saucenao[${hostIndex}] 远程服务器出现问题，请稍后尝试重试`;
                        break;

                    default:
                        logger2.error(`${getTime()},${data}`);
                        //console.error(data);
                        msg = `saucenao[${hostIndex}] ${data.header.message}`;
                        break;
                }
            } else {
                logger2.error(`${getTime()}[error] saucenao[${hostIndex}][data]`);
                logger2.error(data);
                //console.error(`${new Date().toLocaleString()} [error] saucenao[${hostIndex}][data]`);
                //console.error(data);
            }
        })
        .catch(e => {
            logger2.error(`${getTime()}[error] saucenao[${hostIndex}][request]`);
            //console.error(`${new Date().toLocaleString()} [error] saucenao[${hostIndex}][request]`);
            excess = true; //saucenao报错自动使用其它搜索引擎
            if (e.response) {
                if (e.response.status == 429) {
                    msg = `saucenao[${hostIndex}] 搜索次数已达单位时间上限，请稍候再试`;
                } else {
                    logger2.error(`${e.response.data}`);
                    /*console.error(e.response.data);*/
                }
            } else {
                logger2.error(`${e}`);
                /*console.error(e);*/
            }
        });

    if (config.picfinder.debug) {
        logger2.info(`${getTime()}[saucenao][${hostIndex}]\n${msg}`);
        //console.log(`${new Date().toLocaleString()} [saucenao][${hostIndex}]\n${msg}`);
    }

    return {
        success,
        msg,
        warnMsg,
        lowAcc,
        excess,
    };
}

/**
 * 链接混淆
 *
 * @param {string} url
 * @returns
 */
async function confuseURL(url, thumbnail) {
    const {
        hostname
    } = parse(url);
    if (['danbooru.donmai.us', 'konachan.com', 'yande.re'].includes(hostname)) {
        const {
            result,
            path,
            error
        } = await shorten(url);
        return error ? result : `https://j.loli.best/#${path}`;
    }
    return pixivShorten(url, thumbnail);
}

async function getShareText({
    url,
    title,
    thumbnail,
    author_url,
    source
}) {
    let text = `${title}
    ${thumbnail ? CQ.img(thumbnail) : config.picfinder.replys.lowAccImgPlaceholder}
    ${url?await confuseURL(url,thumbnail):""}`;
    if (author_url) text += `\nAuthor: ${await confuseURL(author_url,thumbnail)}`;
    if (source) text += `\nSource: ${await confuseURL(source,thumbnail)}`;
    return text;
}

/**
 * 取得搜图结果
 *
 * @param {string} host 自定义saucenao的host
 * @param {string} imgURL 欲搜索的图片链接
 * @param {number} [db=999] 搜索库
 * @returns Axios对象
 */
function getSearchResult(host, imgURL, db = 999) {
    if (host === 'saucenao.com') host = `https://${host}`;
    else if (!/^https?:\/\//.test(host)) host = `http://${host}`;
    return get(`${host}/search.php`, {
        params: {
            ...saucenaoApiKeyAddition,
            db: db,
            output_type: 2,
            numres: 3,
            url: imgURL,
        },
    });
}

function getTime() {
    return new Date().toLocaleString();
}

export default doSearch;

export {
    snDB
};